packages/discord-request/src/manager.ts

```diff
@@ -0,0 +1,330 @@
+import debug from "debug";
+import { OFFSET, ONE_DAY, ONE_SECOND, sleep } from "./util/time";
+import { getRouteInformation, getRouteKey } from "./util/routes";
+import Queue from "./queue";
+import { RateLimitData, RequestData, RequestOptions, Route } from "./types";
+
+const log = debug("discord-request:manager");
+
+type Bucket = {
+	key: string;
+	lastRequest: number;
+};
+
+export type Config = {
+	api: string;
+	cdn: string;
+	headers: Record<string, string>;
+	retries: number;
+	timeout: number;
+	userAgent: string;
+	version: number;
+	globalRequestsPerSecond: number;
+};
+
+export type Callbacks = {
+	onBucketSweep?: (swept: Map<string, Bucket>) => void;
+	onQueueSweep?: (swept: Map<string, Queue>) => void;
+	onRateLimit?: (data: RateLimitData) => void;
+	onRequest?: (
+		parameters: Route,
+		resource: string,
+		init: RequestInit,
+		retries: number
+	) => void;
+};
+
+type Cache = {
+	shutdownSignal?: AbortSignal;
+	queueSweepInterval?: number;
+	bucketSweepInterval?: number;
+};
+
+export type ManagerArgs = Partial<Config> & Cache & Callbacks;
+
+export class Manager {
+	#token: string | null = null;
+	shutdownSignal?: AbortSignal | null;
+	config: Config;
+
+	globalDelay: Promise<void> | null = null;
+	globalReset = -1;
+	globalRequestCounter: number;
+
+	buckets: Map<string, Bucket>;
+	queues: Map<string, Queue>;
+
+	#bucketSweeper = -1;
+	#queueSweeper = -1;
+
+	bucketSweepInterval: number;
+	queueSweepInterval: number;
+
+	onBucketSweep?: Callbacks["onBucketSweep"];
+	onQueueSweep?: Callbacks["onQueueSweep"];
+	onRateLimit?: Callbacks["onRateLimit"];
+	onRequest?: Callbacks["onRequest"];
+
+	constructor({
+		// Request Config
+		api,
+		version,
+		cdn,
+		headers,
+		userAgent,
+		retries,
+		timeout,
+		globalRequestsPerSecond,
+		// Manager Options
+		shutdownSignal,
+		onBucketSweep,
+		onQueueSweep,
+		bucketSweepInterval,
+		queueSweepInterval,
+		onRateLimit,
+		onRequest,
+	}: ManagerArgs) {
+		this.config = {
+			api: api ?? "https://discord.com/api",
+			version: version ?? 10,
+			cdn: cdn ?? "https://cdn.discordapp.com",
+			headers: headers ?? {},
+			userAgent: userAgent ?? `Discord Request v0`,
+			retries: retries ?? 3,
+			timeout: timeout ?? 15 * ONE_SECOND,
+			globalRequestsPerSecond: globalRequestsPerSecond ?? 50,
+		};
+
+		this.buckets = new Map();
+		this.queues = new Map();
+
+		this.globalRequestCounter = this.config.globalRequestsPerSecond;
+
+		this.bucketSweepInterval = bucketSweepInterval ?? 0;
+		this.queueSweepInterval = queueSweepInterval ?? 0;
+		this.onBucketSweep = onBucketSweep;
+		this.onQueueSweep = onQueueSweep;
+		this.onRateLimit = onRateLimit;
+		this.onRequest = onRequest;
+
+		this.shutdownSignal = shutdownSignal;
+		this.shutdownSignal?.addEventListener("abort", this.shutdown);
```

Comment:
```suggestion
		this.shutdownSignal?.addEventListener("abort", this.shutdown.bind(this));
```

pretty sure this loses the `this` context when you pass it without a bind or explicit call